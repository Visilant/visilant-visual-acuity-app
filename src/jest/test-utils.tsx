import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { fireEvent, render, RenderOptions, screen } from '@testing-library/react-native'
import { WrapperProps } from './types'
import { ThemeProvider } from 'styled-components/native'
import { appTheme } from '@components/theme/theme'
import { LocalizationProvider } from '@localizations/Provider'
import { store } from '@store'

const Providers: React.FunctionComponent<WrapperProps> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <ThemeProvider theme={appTheme}>
        <LocalizationProvider>{children}</LocalizationProvider>
      </ThemeProvider>
    </ReduxProvider>
  )
}

const customRender = <T,>(
  ui: React.ReactElement<unknown, string | React.JSXElementConstructor<T>>,
  options?: RenderOptions
) => {
  return render(ui, {
    wrapper: Providers,
    ...options
  })
}

export { fireEvent, customRender as render, screen }
