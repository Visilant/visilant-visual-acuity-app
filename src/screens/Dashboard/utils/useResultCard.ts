import { Examination } from '@shared/domain/examination'
import dayjs from 'dayjs'

export const useResultCard = (examination: Examination) => {
  return {
    patientName: examination.patientName,
    date: dayjs(examination.mobileCreatedAt).format('YYYY. MM. DD. HH:mm'),
    right: examination.normal?.right,
    left: examination.normal?.left,
    rightPinhole: examination.pinhole?.right,
    leftPinhole: examination.pinhole?.left
  }
}
