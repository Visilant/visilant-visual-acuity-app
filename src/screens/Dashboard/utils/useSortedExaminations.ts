import { useExamination } from '@hooks/useExamination'
import { useMemo } from 'react'

export const sortByMobileCreatedAt = (a, b) =>
  b.mobileCreatedAt.getTime() - a.mobileCreatedAt.getTime()

export const useSortedExaminations = () => {
  const examinations = useExamination()
  const sortedExaminations = useMemo(
    () => examinations.values.sort(sortByMobileCreatedAt),
    [examinations]
  )
  return sortedExaminations
}
