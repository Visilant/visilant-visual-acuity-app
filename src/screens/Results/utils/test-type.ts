export enum TestType {
  Visual = 'results.visual',
  Pinhole = 'results.pinhole'
}
