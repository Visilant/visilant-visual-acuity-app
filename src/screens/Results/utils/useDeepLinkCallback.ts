import { AppConfig } from '@config'
import { Examination } from '@shared/domain/examination'
import { ExaminationMapper } from '@shared/mappers/examination-mapper'
import { useAppDispatch } from '@store'
import { clearDeepLinkParams } from '@store/examinations'
import { DeepLinkParams } from '@store/examinations/types'
import { useCallback } from 'react'
import { Linking } from 'react-native'

const createUrl = (baseUrl: string, callbackId: string, examinationString) =>
  `${baseUrl}/${callbackId}/${encodeURIComponent(examinationString)}`

export const useDeepLinkCallback = () => {
  const dispatch = useAppDispatch()

  const navigateToTelemed = useCallback(
    async (examination: Examination, deepLinkParams: DeepLinkParams) => {
      const examinationString = ExaminationMapper.fromDomainToSerializedResults(examination)
      const { deepLinkBaseUrl } = AppConfig.getConfig()
      const url = createUrl(deepLinkBaseUrl, deepLinkParams.callbackId, examinationString)
      dispatch(clearDeepLinkParams())
      if (await Linking.canOpenURL(url)) {
        Linking.openURL(url)
      }
    },
    [dispatch]
  )
  return {
    navigateToTelemed
  }
}
