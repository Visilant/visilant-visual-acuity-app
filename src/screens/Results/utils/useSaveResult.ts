import { AlertType } from '@components/Snackbar/types'
import { useStackNavigation } from '@screens/stack'
import { useSnackbar } from '@hooks/useSnackbar'
import { useAppDispatch, useAppSelector } from '@store'
import { clearActiveResults, saveResult, selectDeepLinkParams } from '@store/examinations'
import { Examination } from '@shared/domain/examination'
import { useCallback } from 'react'
import { useDeepLinkCallback } from './useDeepLinkCallback'
import { ExaminationToSave } from '@shared/domain/examination-type'

export const snackbarFeedback = {
  success: { type: AlertType.success, text: 'success', namespace: 'alerts' }
}

export const useSaveResult = () => {
  const navigation = useStackNavigation()
  const snackbar = useSnackbar()
  const dispatch = useAppDispatch()
  const deepLinkParams = useAppSelector(selectDeepLinkParams)
  const { navigateToTelemed } = useDeepLinkCallback()

  const save = useCallback(
    async (results: ExaminationToSave) => {
      dispatch(clearActiveResults())
      if (deepLinkParams === undefined) {
        dispatch(saveResult(results))
      } else {
        const { patientName, comment, normal, pinhole } = results
        const examination = Examination.fromResults({ patientName, comment, normal, pinhole })
        navigateToTelemed(examination, deepLinkParams)
      }
      snackbar.addAlert(snackbarFeedback.success)
      navigation.navigate('Dashboard')
    },
    [deepLinkParams, navigation, snackbar, dispatch]
  )

  return { save }
}
