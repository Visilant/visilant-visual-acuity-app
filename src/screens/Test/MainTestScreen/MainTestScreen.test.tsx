import { render } from '@jest/test-utils'
import { ScreenState } from '../utils/screen-state'
import '@testing-library/jest-native'
import MainTestScreen from './MainTestScreen'

const mockState = ScreenState.testing

const mockNavigate = jest.fn()
jest.mock('@screens/stack', () => ({
  useStackNavigation: () => ({ navigate: mockNavigate }),
  useStackRoute: () => ({ params: {} })
}))
jest.mock('../utils/useMainTestScreenRotation', () => ({
  useMainTestScreenRotation: () => ({
    screenState: mockState
  })
}))

jest.mock('../utils/useSetNavigationOrientation', () => ({
  useSetNavigationOrientation: () => {}
}))

jest.mock('@screens/Test/components/TestingFlow', () => ({
  TestingFlow: () => <>TestingFlowMock</>
}))

describe('Main test screen', () => {
  // todo: test orientation screen
  test('render testing flow', async () => {
    const mainTestScreen = render(<MainTestScreen />)

    expect(mainTestScreen).toMatchSnapshot()
  })
})
