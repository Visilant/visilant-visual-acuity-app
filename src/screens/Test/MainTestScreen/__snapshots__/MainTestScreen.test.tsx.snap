// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main test screen render testing flow 1`] = `
<View
  style={
    {
      "backgroundColor": "#EDF1F4",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "gap": 30,
      "justifyContent": "center",
      "paddingLeft": 30,
      "paddingRight": 30,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <Image
      source={1}
      style={
        {
          "marginLeft": -30,
        }
      }
    />
  </View>
  <Text
    style={
      [
        {
          "color": "#23255A",
          "fontFamily": "Caprasimo",
          "fontSize": 36,
          "lineHeight": 38,
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "textAlign": "center",
        },
        "
  ",
        "
  ",
        [Function],
        "
  font-family: Caprasimo;
  font-size: 36px;
  line-height: 38px;
",
        "
  text-align: center;
",
      ]
    }
  >
    Something went wrong
  </Text>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      size={1}
      style={
        {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "#EF4823",
          "borderBottomLeftRadius": 31,
          "borderBottomRightRadius": 31,
          "borderTopLeftRadius": 31,
          "borderTopRightRadius": 31,
          "flexWrap": "nowrap",
          "height": 62,
          "justifyContent": "center",
          "overflow": "hidden",
          "paddingBottom": 12,
          "paddingLeft": 30,
          "paddingRight": 30,
          "paddingTop": 12,
        }
      }
      type="normal"
    >
      <View
        style={{}}
      />
      <Text
        props={
          {
            "size": 1,
            "type": "normal",
          }
        }
        style={
          [
            {
              "color": "#E6EEF4",
              "flexWrap": "nowrap",
              "fontFamily": "Libre-Franklin-Bold",
              "fontSize": 24,
              "lineHeight": 38,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
              "position": "relative",
              "textAlign": "center",
              "top": -1,
            },
            "
  position: relative;
  top: ",
            [Function],
            ";
  color: ",
            [Function],
            ";
  font-family: Libre-Franklin-Bold;

  ",
            [Function],
            ";
  text-align: center;
  flex-wrap: nowrap;
",
          ]
        }
      >
        TRY AGAIN...
      </Text>
    </View>
  </View>
</View>
`;
